explainability-service/src/main/java/org/kie/trustyai/service/data/storage/BatchReader.java

```diff
@@ -0,0 +1,52 @@
+package org.kie.trustyai.service.data.storage;
+
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Scanner;
+
+import org.apache.commons.collections4.queue.CircularFifoQueue;
+
+public class BatchReader {
+
+    private BatchReader() {
+    }
+
+    public static List<String> readEntries(InputStream stream, int batchSize) throws IOException {
+        final CircularFifoQueue<String> queue = new CircularFifoQueue<>(batchSize);
+        String header = "";
+        try (Scanner sc = new Scanner(stream, StandardCharsets.UTF_8)) {
+            boolean capturedHeader = false;
+            while (sc.hasNextLine()) {
+                String line = sc.nextLine();
+                if (capturedHeader) {
+                    queue.add(line);
+                } else {
+                    header = line;
+                    capturedHeader = true;
+                }
+
+            }
+            if (sc.ioException() != null) {
+                throw sc.ioException();
+            }
+        }
+
+        final List<String> entries = new ArrayList<>();
+        entries.add(header);
+        entries.addAll(new ArrayList<>(queue));
```

Comment:
Isn't the queue just going to be full of the last $batchSize entries of the inputstream at this point, not the actual batch we're trying to get?

Reply:
@RobGeada as discussed offline, the batch reader is reading the last lines (most recent) since the most recent data is appended to the end. So effectively this is returning a `header` + `tail` (of size `batchSize`).